From: SeongJae Park <sj@kernel.org>
Date: Wed, 14 Jan 2026 07:13:19 -0800
Subject: [PATCH] mm/damon/tests/core-kunit: add a test for
 damon_is_last_region()

There was a bug in the function.  Add a kunit test to not reintroduce
the bug.

The issue report:
https://lore.kernel.org/20260114152049.99727-1-sj@kernel.org/

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/tests/core-kunit.h | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/mm/damon/tests/core-kunit.h b/mm/damon/tests/core-kunit.h
index 547985dbf64cf..982edde8a7549 100644
--- a/mm/damon/tests/core-kunit.h
+++ b/mm/damon/tests/core-kunit.h
@@ -1270,6 +1270,28 @@ static void damon_test_get_apply_max_region_sz(struct kunit *test)
 	KUNIT_EXPECT_EQ(test, damon_nr_regions(t), 10);
 }
 
+static void damon_test_is_last_region(struct kunit *test)
+{
+	struct damon_region *r;
+	struct damon_target *t;
+	int i;
+
+	t = damon_new_target();
+	if (!t)
+		kunit_skip(test, "target alloc fail\n");
+
+	for (i = 0; i < 4; i ++) {
+		r = damon_new_region(i * 2, (i + 1) * 2);
+		if (!r) {
+			damon_free_target(t);
+			kunit_skip(test, "region alloc %d fail\n", i);
+		}
+		damon_add_region(r, t);
+		KUNIT_EXPECT_TRUE(test, damon_is_last_region(r, t));
+	}
+	damon_free_target(t);
+}
+
 static struct kunit_case damon_test_cases[] = {
 	KUNIT_CASE(damon_test_target),
 	KUNIT_CASE(damon_test_regions),
@@ -1297,6 +1319,7 @@ static struct kunit_case damon_test_cases[] = {
 	KUNIT_CASE(damon_test_feed_loop_next_input),
 	KUNIT_CASE(damon_test_set_filters_default_reject),
 	KUNIT_CASE(damon_test_get_apply_max_region_sz),
+	KUNIT_CASE(damon_test_is_last_region),
 	{},
 };
 
-- 
2.47.3

